### PR TITLE
Fix synchronous relationship parsing and serializing

### PR DESCRIPTION
**Pull request by delkopiso**
# What's up
Right now, when you have a model structure like:
```javascript
User = DS.Model.extend({
  name: DS.attr('string'),
  posts: DS.hasMany('post', { async: false })
});

Post = DS.Model.extend({
  title: DS.attr('string'),
  body: DS.attr('string'),
  comments: DS.hasMany('comment', { async: false })
});

Comment = DS.Model.extend({
  body: DS.attr('string')
});
```
The generated query when you run `this.store.findRecord('user', 1)` looks like:
```
query user(id: "1') { name posts { title body } }
```

This is not an accurate representation of what the query should look like from the model descriptions.

Also when a payload is deeply nested like:
```json
{
  "data": {
    "user": {
      "id": "1",
      "name": "John",
      "posts": [{
        "id": "1",
        "title": "first post title",
        "body": "first post body",
        "comments": [{
          "id": "1",
          "body": "first post first comment"
        }, {
          "id": "2",
          "body": "first post second comment"
        }]
      }, {
        "id": "2",
        "title": "second post title",
        "body": "second post body",
        "comments": [{
          "id": "3",
          "body": "second post first comment"
        }, {
          "id": "4",
          "body": "second post second comment"
        }]
      }]
    }
  }
}
```
The serializer generates the following JSONAPI document:
```json
{
  "data": {
    "type": "user",
    "id": "1",
    "attributes": {
      "name": "John"
    },
    "relationships": {
      "posts": {
        "data": [{
          "type": "post",
          "id": "1"
        }, {
          "type": "post",
          "id": "2"
        }]
      }
    }
  },
  "included": [{
    "type": "post",
    "id": "1",
    "attributes": {
      "title": "first post title",
      "body": "first post body"
    },
    "relationships": {}
  }, {
    "type": "post",
    "id": "2",
    "attributes": {
      "title": "second post title",
      "body": "second post body"
    },
    "relationships": {}
  }]
}
``` 
Which, again, is not in line with what the model description is aiming for.

# What this does
- Makes the parser goes through more than just the top level relationships and attributes. This means the generated query for the aforementioned model description will look like:
```
query user(id: "1") { name posts { title body comments { body } } }
```

- Updates serializer to deal with deeply nested payloads. So when you receive data as described above, the serializer generates the following JSONAPI document:
```json
{
  "data": {
    "type": "user",
    "id": "1",
    "attributes": {
      "name": "John"
    },
    "relationships": {
      "posts": {
        "data": [{
          "type": "post",
          "id": "1"
        }, {
          "type": "post",
          "id": "2"
        }]
      }
    }
  },
  "included": [{
    "type": "post",
    "id": "1",
    "attributes": {
      "title": "first post title",
      "body": "first post body"
    },
    "relationships": {
      "comments": {
        "data": [
          {"type": "comment", "id": "1"},
          {"type": "comment", "id": "2"},
        ]
      }
    }
  }, {
    "type": "post",
    "id": "2",
    "attributes": {
      "title": "second post title",
      "body": "second post body"
    },
    "relationships": {
      "comments": {
        "data": [
          {"type": "comment", "id": "3"},
          {"type": "comment", "id": "4"},
        ]
      }
    }
  }, {
    "type": "comment",
    "id": "1",
    "attributes": {
      "body": "first post first comment"
    },
    "relationships": {}
  }, {
    "type": "comment",
    "id": "2",
    "attributes": {
      "body": "first post second comment"
    },
    "relationships": {}
  }, {
    "type": "comment",
    "id": "3",
    "attributes": {
      "body": "second post first comment"
    },
    "relationships": {}
  }, {
    "type": "comment",
    "id": "4",
    "attributes": {
      "body": "second post second comment"
    },
    "relationships": {}
  }]
}
```

### CODE REVIEW

**Comment by toreriklinnerud** [comment]
💯

**Comment by delkopiso** [comment]
@Arkham review?

File: tests/unit/serializer-test.js [hunk 1]
```diff
@@ -197,79 +234,100 @@ test('normalizing with synchronous relationships', function(assert) {
...
-    longBody: 'The long body'
-  };
-
-  let projectRels = {
-    author: { kind: 'belongsTo', key: 'author', data: { id: '1', modelName: 'user' } },
-    posts: { kind: 'hasMany', key: 'posts', data: [{ id: '1', modelName: 'post' }] },
-  };
-
-  assert.deepEqual(serializer.serialize(new SnapshotDouble('project', projectAttrs, projectRels)), expected);
-});
```
Comment (Arkham) [inline tests/unit/serializer-test.js]: I don't see that much value in these unit tests now that we have proper integration tests. What do you think about removing them?

**Comment by Arkham** [comment]
This looks great :shipit:

File: tests/unit/serializer-test.js (hunk 1 above)
Comment (Arkham) [inline tests/unit/serializer-test.js]: This would also allow us to remove all those iffy helpers who stub out Ember's container and so on :)

**Pull request merged**